File: 313. Super Ugly Number.py
# timeout

import sys
class Solution(object):
    def nthSuperUglyNumber(self, n, primes):
        """
        :type n: int
        :type primes: List[int]
        :rtype: int
        """
        inds = [0 for i in xrange(len(primes))]
        res = [1]
        dic = {}
        while len(res) < n:
            minn = sys.maxint
            index = 0
            for i in xrange(len(primes)):
                temp = primes[i]*res[inds[i]]
                if temp < minn:
                    minn = temp
                    index = i
            inds[index] += 1
            if minn not in dic:
                res += [minn]
                dic[minn] = 1
        return res[n-1]

File: 313. Super Ugly Number.cpp
class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {
        vector <long long int> candidates (primes.size(), 1);
        vector <int> inds (primes.size());
        vector <long long int> res {1};
        long long int minn = 1;
        while (res.size() < n)
        {
            for (int i = 0; i < primes.size(); i++)
            {
                if (minn == candidates[i])
                {
                    candidates[i] = primes[i] * res[inds[i]];
                    inds[i] ++;
                }
            }
            int temp = INT_MAX;
            for (int j = 0; j < primes.size(); j++)
            {
                if (candidates[j] < temp)
                    temp = candidates[j];
            }
            minn = temp;
            res.push_back(minn);
        }
        return res[n - 1];
    }
};
